proposals/NNNN-constant-buffers.md

```diff
@@ -0,0 +1,132 @@
+
+At the same time Clang needs to recognize this class represents a constant buffer and the contained type fields are accessed using the `.` operator on `ConstantBuffer` instance. In other words treating `ConstantBuffer<MyConstants> CB;` as if it was declared as `cbuffer __CB { MyConstants CB; }`. The exact way how to do this is TBD.
+
+### Lowering Constant Buffers to LLVM IR
+
+During CodeGen constant buffers will be lowered to global variables with LLVM target type `target("dx.CBuffer", ..)` which will include type information about constants, the buffer size and its memory layout.
+
+Note: LLVM target types can optionally include a list of one or more types and a list of one or more integer constants. We can use these lists to encode any information needed for lowering from LLVM IR to DXIL and SPIRV.
+
+To encode the shape of the constant buffer the LLVM target type will include a structure type that represents the constant buffer variable declarations.
+
+The size of the constant buffer will be included as the first item in the list of integer constants. The rest of the list will be used to encode the constant buffer layout. The layout will always be included whether the constant buffer uses any `packoffset` attributes or not. The exact way how the layout will be encoded is TBD and will be covered in a separate design document.
+
+For simplicity, let's assume the layout will be encoded as a list of offsets of all cbuffer declarations. In that case this example:
+
+```c++
+cbuffer MyConstants {
+  float2 a;
+  float b[2];
+  int c;
+}
+```
+
+Would be lowered to LLVM target type:
+
+```
+@MyConstants.cb = global target("dx.CBuffer", { <2 x float>, [2 x float], i32}, 40, 0, 16, 32, 36)
+```
+
+This layout encoding can obviously get very long and unwieldy for more complicated cbuffers, and especially since target type parameters are all included in the name mangling for function overloads. We need to investigate how to make it smaller, or at least more manageable.
+
+One possibility is compressing the list of offsets into a smaller number of integers - taking advantage of the fact that outside of `packoffset` use the difference between two adjancent offsets is never more than 16. The compression could also include repetition construct that would help with encoding of array offset. But, as with any compressions, there's always the chance of degenerate cases that will end up with the compressed shape being the same size or larger than the original.
+
+> Note: The most tricky part of the layout encoding are probably arrays of structures because the structure-specific layout gets repeated many times and might not be easy to compress. One idea how to solve this could be translating structures embedded in `cbuffer` into separate target types with their own encoded layout and including them in the `cbuffer` target type. It is not clear though if this is possible (probably yes) or if it would actually make things easier or not.
```

Comment:
I don't think we should be encoding the offset of each element of arrays, nor should we be encoding the offsets of inner struct members for each instance of them in the outer struct.  I think we need to associate type information, containing member offsets, with struct types, like we use for type annotations in DXIL.  This is a lot more efficient, only encoding the offsets we need once.  Then we need some way to connect the target type with the annotated type, or the annotation metadata itself.  Perhaps we can have cbuffer type metadata which references the target type and the contained type, then lists member info, including these offsets.  Other member info will be useful/necessary, such as matrix identification and orientation information.